projects/sampling/executor.py

```diff
@@ -0,0 +1,185 @@
+
+
+class Executor(base_executor.BaseExecutor):
+  """Executor for Undersampler."""
+
+  def Do(
+    self,
+    input_dict: Dict[Text, List[types.Artifact]],
+    output_dict: Dict[Text, List[types.Artifact]],
+    exec_properties: Dict[Text, Any],
+  ) -> None:
+
+    """Undersampler executor entrypoint.
+
+    Args:
+      input_dict: Input dict from input key to a list of artifacts, including:
+      - examples: A list of type `standard_artifacts.Examples` which should
+      contain custom splits specified in splits_config. If custom split is
+      not provided, this should contain two splits 'train' and 'eval'.
+      output_dict: Output dict from key to a list of artifacts, including:
+      - undersampled_examples: Undersamplerd examples, only for the given
+      splits as specified in splits. May also include copies of the
+      other non-undersampled spits, as specified by keep_classes.
+      exec_properties: A dict of execution properties, including:
+      - name: Optional unique name. Necessary if multiple components are
+      declared in the same pipeline.
+      - label: The name of the column containing class names to undersample by.
+      - splits: A list containing splits to undersample. Defaults to ['train'].
+      - copy_others: Determines whether we copy over the splits that aren't
+      undersampled, or just exclude them from the output artifact. Defualts
+      to True.
+      - shards: The number of files that each undersampled split should
+      contain. Default 0 is Beam's tfrecordio function's default.
+      - keep_classes: A list determining which classes that we should
+      not undersample. Defaults to None.
+
+    Returns:
+      None
+    """
+
+    self._log_startup(input_dict, output_dict, exec_properties)
+    label = exec_properties[component.UNDERSAMPLER_LABEL_KEY]
+    splits = json_utils.loads(exec_properties[component.UNDERSAMPLER_SPLIT_KEY])
+    copy_others = exec_properties[component.UNDERSAMPLER_COPY_KEY]
+    shards = exec_properties[component.UNDERSAMPLER_SHARDS_KEY]
+    keep_classes = json_utils.loads(exec_properties[component.UNDERSAMPLER_CLASSES_KEY])
+
+    input_artifact = artifact_utils.get_single_instance(input_dict[component.UNDERSAMPLER_INPUT_KEY])
+    output_artifact = artifact_utils.get_single_instance(output_dict[component.UNDERSAMPLER_OUTPUT_KEY])
+
+    if copy_others:
+      output_artifact.split_names = input_artifact.split_names
+    else:
+      output_artifact.split_names = artifact_utils.encode_split_names(splits)
+
+    # Fetch the input uri for each split
+    split_data = {}
+    for split in artifact_utils.decode_split_names(input_artifact.split_names):
+      uri = artifact_utils.get_split_uri([input_artifact], split)
+      split_data[split] = uri
+
+    for split, uri in split_data.items():
+      if split in splits:  # Undersampling split
+        output_dir = artifact_utils.get_split_uri([output_artifact], split)
+        os.mkdir(output_dir)
```

Comment:
`os.mkdir` will only work if your metadata is stored locally, if you're using gcs as a metadata bucket (which I am) this won't do anything. I'm not sure if you even need to `mkdir` here since I think all directories might be created automatically when you write your first file. I may be wrong here. Alternatively there's `tf.io.gfile.mkdir()` which I think will create a directory irrespective of the file system.